zkvm/src/constraints.rs

```diff
@@ -89,21 +95,60 @@ impl Constraint {
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), VMError> {
         match self {
-            Constraint::Eq(expr1, expr2) => expr1.to_r1cs_lc() - expr2.to_r1cs_lc(),
+            Constraint::Eq(expr1, expr2) => {
+                let lc = expr1.to_r1cs_lc() - expr2.to_r1cs_lc();
+                match (expr1.to_scalar(), expr2.to_scalar()) {
+                    (Some(x), Some(y)) => Ok((lc, Some(x - y))),
+                    (_, _) => Ok((lc, None)),
+                }
+            }
             Constraint::And(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 let z = cs.challenge_scalar(b"ZkVM.verify.and-challenge");
-                a + z * b
+                match (a_assg, b_assg) {
+                    (Some(a_assg), Some(b_assg)) => Ok((a + z * b, Some(a_assg + z * b_assg))),
+                    (_, _) => Ok((a + z * b, None)),
+                }
             }
             Constraint::Or(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 // output expression: a * b
                 let (_l, _r, o) = cs.multiply(a, b);
-                r1cs::LinearCombination::from(o)
+                match (a_assg, b_assg) {
+                    (Some(a_assg), Some(b_assg)) => {
+                        Ok((r1cs::LinearCombination::from(o), Some(a_assg * b_assg)))
+                    }
+                    (_, _) => Ok((r1cs::LinearCombination::from(o), None)),
+                }
+            }
+            Constraint::Not(c1) => {
+                let (x, x_assg) = c1.flatten(cs)?;
+                match x_assg {
+                    Some(x_assg) => {
+                        let is_zero = (x_assg == Scalar::zero()) as u8;
+                        let y_assg = Scalar::conditional_select(
+                            &Scalar::zero(),
+                            &Scalar::one(),
+                            is_zero.into(),
+                        );
+                        let w = Scalar::conditional_select(&x_assg, &Scalar::one(), is_zero.into());
+                        let w = w.invert();
+
+                        let (x, y, o) = cs.multiply(x, y_assg.into());
+                        cs.constrain(o.into());
+                        let (_x, _w, xw) = cs.multiply(x.into(), w.into());
+                        cs.constrain(xw - Scalar::one() + y);
+                        Ok((r1cs::LinearCombination::from(y), Some(y_assg)))
+                    }
+                    None => Err(VMError::WitnessMissing),
```

Comment:
if the x assignment is missing, then we are inside the verifier, so we simply propagate None to all other assignments. Therefore, this whole method can remain being non-failable. 

As of `y_assg`, we should compute it only if x is Some, but keep it None if x is unassigned.